multigen/pipes.py

```diff
@@ -252,6 +238,42 @@ def get_config(self):
         cfg.update(self.pipe_params)
         return cfg
 
+    def gen(self, inputs):
+        inputs = {**inputs}
+        inputs.update(self.pipe_params)
+        inputs.update({"image": self._condition_image})
+        image = self.pipe(**inputs).images[0]
+        return image
+
+
+class CIm2ImPipe(Cond2ImPipe):
+
+    def __init__(self, model_id, ctypes=["soft"], **args):
+        super().__init__(model_id, ctypes, **args)
+        # The difference from Cond2ImPipe is that the conditional image is not
```

Comment:
From comment it looks like CIm2ImPipe is more general than Cond2ImPipe and Cond2ImPipe only handles the case when we don't process conditioning image. We can merge these cases into a single class with additional "if" for empty ctypes

Reply:
No. `CIm2Im` accepts an ordinary image and turns it into a conditioning image. It is not about empty `ctypes`, and it is not about generality. It is possible to merge these two classes into one with an additional flag, which would indicate whether to process the input image or take it as a ready-to-use conditioning image, but I don't see the need for this. `CIm2Im` can be less general supporting less number of conditioning types. There can be also other special types of conditioning built on top of the base `Cond2Im` class (e.g. panorama generation). If we put everything in one class with flag-based switching, it will be very difficult to read.

Reply:
Well, it's good to split big classes, but it looks like CIm2ImPipe can be parent and Cond2ImPipe be subclass. It just need to set self.ctypes = ['identity']. _proc_cimg then will return original image, it handles this condition at line 310.

```
class Cond2ImPipe(CIm2ImPipe):
    def __init__(self, model_id, **args):
          super().__init__(model_id, ctypes=['identity'], **args)
```

Reply:
I thought about `'identity'`, but your approach doesn't work, because we need to load control nets corresponding to `ctypes` (as well as control strengths).